src/protocol/sort/generate_sort_permutation.rs

```diff
@@ -0,0 +1,159 @@
+
+/// This is an implementation of `GenPerm` (Algorithm 6) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+#[derive(Debug)]
+pub struct GenerateSortPermutation<'a> {
+    input: &'a [(u64, u64)],
+    num_bits: u8,
+}
+
+impl<'a> GenerateSortPermutation<'a> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a [(u64, u64)], num_bits: u8) -> GenerateSortPermutation {
+        Self { input, num_bits }
+    }
+
+    #[allow(dead_code)]
+    #[embed_doc_image("semi_honest_sort", "images/sort/semi-honest-sort.png")]
+    /// This protocol generates permutation of a stable sort for the given shares of inputs.
+    /// ![Generate sort permutation steps][semi_honest_sort]
+    /// Steps
+    /// For the 0th bit
+    /// 1. Get replicated shares in Field using modulus conversion
+    /// 2. Compute bit permutation that sorts 0th bit
+    /// For 1st to N-1th bit of input share
+    /// 1. Get replicated shares in Field using modulus conversion
+    /// 2. Sort ith bit based on i-1th bits by applying i-1th composition on ith bit
+    /// 3  Compute bit permutation that sorts ith bit
+    /// 4. Compute ith composition by composing i-1th composition on ith permutation
+    /// In the end, n-1th composition is returned. This is the permutation which sorts the inputs
+    pub async fn execute<F: Field>(
+        &self,
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let ctx_0 = ctx.narrow(&Sort(0));
+        let bit_0 = convert_shares_for_a_bit(
+            ctx_0.narrow(&ModulusConversion),
+            self.input,
+            self.num_bits,
+            0,
+        )
+        .await?;
+        let bit_0_permutation = BitPermutation::new(&bit_0)
+            .execute(ctx_0.narrow(&BitPermutationStep))
+            .await?;
+
+        let mut composed_permutation = bit_0_permutation;
+        for bit_num in 1..self.num_bits {
+            let ctx_bit = ctx.narrow(&Sort(bit_num));
+            let bit_i = convert_shares_for_a_bit(
+                ctx_bit.narrow(&ModulusConversion),
+                self.input,
+                self.num_bits,
+                bit_num,
+            )
+            .await?;
+            let applied_bit = SecureApplyInv::execute(
+                ctx_bit.narrow(&ApplyInv),
+                bit_i,
+                composed_permutation.clone(),
+            )
+            .await?;
+
+            let bit_i_permutation = BitPermutation::new(&applied_bit)
+                .execute(ctx_bit.narrow(&BitPermutationStep))
+                .await?;
+
+            let new_composed_permutation = Compose::new(composed_permutation, bit_i_permutation)
+                .execute(ctx_bit.narrow(&ComposeStep))
+                .await?;
+
+            composed_permutation = new_composed_permutation;
```

Comment:
```
composed_permutation = Compose::new(composed_permutation, bit_i_permutation)
    .execute(ctx_bit.narrow(&ComposeStep))
    .await?;
```

Reply:
trying to keep new composed different from older one, so renaming composed_less_significant_bits_permutation and composed_i_permutation